---
output:
    pdf_document:
        toc: true
        toc_depth: 6
version: 0.1.1
author: Tamas Ryszard Sztanka-Toth, Nikolaos Karaiskos
email: tamasryszard.sztanka-toth@mdc-berlin.de, nikolaos.karaiskos@mdc.berlin.de
license: GPL
---

```{r knitr_options, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
  cache = F,
  autodep = TRUE,
  message = FALSE,
  warning = FALSE,
  comment = NA
)
```

```{r load_libraries, echo =F}
library(tidyverse)
library(yaml)
library(cowplot)
library(knitr)

theme_set(theme_cowplot(12))
#snakemake = read_rds('/scratch/home/tsztank/snakemake.rds')
```

```{r functions, echo = F}
readStarLog <- function(log_file){
		out = list()
		lines = readLines(log_file)
	
		out$input_reads = (lines[6] %>% strsplit('\t') %>% unlist)[2] %>% as.integer

		out$uniq_mapped_reads = (lines[9] %>% strsplit('\t') %>% unlist)[2] %>% as.integer

		#out$avg_length = (lines[11] %>% strsplit('\t') %>% unlist)[2] %>% as.numeric
		
        tibble(observation=names(out), value=unlist(unname(out)))
	}
```


```{r read_qc_params_file, echo = F}
qc_params <- read_yaml(snakemake@input[['parameters_file']])
```

#### Run information

- **Project**: `r qc_params$project_id`
- **Sample**: `r qc_params$sample_id`
- **Puck**: `r qc_params$puck_id`
- **Experiment**: `r qc_params$experiment`
- **Sequencing date**: `r qc_params$sequencing_date`

saturation analysis v.0.1.1, generated on `r format(Sys.time(), '%d/%B/%Y')`

contact: tamasryszard.sztanka-toth@mdc-berlin.de, nikolaos.karaiskos@mdc-berlin.de

#### Downstream stats

```{r parse_metadata, echo = F}
metadata <- readStarLog(snakemake@input$star_log) %>%
    rbind(read_table2(snakemake@input$reads_type_out, col_names=c('observation', 'value'))) %>%
    # convert to millions
    mutate(value = round(value / 1e6, 2)) %>%
    {
        mutate(., input_reads = filter(., observation == 'input_reads')$value)
    } %>%
    mutate(label = ifelse(observation == 'input_reads', value, paste0(value, ' (', round(value/input_reads*100, 1), '%)'))) %>%
    dplyr::select(observation, label) %>%
    spread(observation, label) %>%
    dplyr::rename(as.utr = UTR,
                  intronic = INTRONIC,
                  intergenic = INTERGENIC,
                  as.cds = CODING) %>%
    # reorder columns
    dplyr::select(input_reads, uniq_mapped_reads, intergenic, intronic, as.cds, as.utr) %>%
    kable

metadata
```

#### Saturation analysis

In order to know whether we would gain more from sequencing deeper, we downsampled the data (the final.bam file) to contain 10%, 20%... 90% reads, and then we created the DigitalExpression matrix (as in the normal dropseq pipeline).

This can give us insight, whether we have reached the saturation point (in terms of median umi per cell and median genes per cell) or whether we should sequence deeper.

Results of this are plotted below.

```{r read_summaries, echo = F}
downsampled_summaries <- snakemake@input[startsWith(names(snakemake@input), 'downsampled')]

dge_data <- tibble(percentage = names(downsampled_summaries),
                   filename = unlist(downsampled_summaries)) %>%
            mutate(content = map(filename, 
                                 ~ read_table2(., skip=6))) %>%
            unnest(content) %>%
            select(-filename) %>%
            separate(percentage, into = c(NA, 'percentage')) %>%
            mutate(pcr = NUM_GENIC_READS / NUM_TRANSCRIPTS) %>%
            rename(umis = NUM_TRANSCRIPTS,
                   genes = NUM_GENES,
                   reads = NUM_GENIC_READS)
```

\newpage 

##### we bin the data into percentalise by reads per bead

```{r plot_by_deciles, echo = F, fig.height=7, fig.width=7}
decile_dat <- dge_data %>%
    group_by(percentage) %>%
    mutate(cumsum_reads = cumsum(reads),
           decile_limit = sum(reads)/10,
           # put beads into deciles by number of reads
           decile = floor(cumsum_reads / decile_limit) + 1) %>%
    # get top 10 deciles, 11 is an artifact of rounding, last beads
    filter(decile < 11) %>%
    group_by(percentage, decile) %>%
    summarise(median_reads = median(reads),
              median_genes = median(genes),
              median_pcr = median(pcr),
              median_umis = median(umis),
              num_beads = n()) %>%
    gather('observation', 'value', median_reads:num_beads) %>%
    mutate(decile = factor(decile), 
           percentage = as.integer(percentage),
           observation = factor(observation, levels = c(
                      'median_reads', 
                      'median_umis', 
                      'median_genes',
                      'median_pcr', 'num_beads')))

decile_dat %>%
    ggplot(aes(percentage, value, color= decile, fill = decile)) +
        geom_smooth(formula = y~log(x), size = 0.6) +
        geom_point(size=2,  color = 'black', pch=21) + 
        scale_x_continuous(breaks=seq(0, 100, 10)) +
        facet_wrap(~observation, scales = 'free_y', ncol=2) +
        labs(y='', x='downsampling percentage')
```


\newpage

```{r plot_observations, echo = F}
plot_data <- function(umi_cutoff=1){
    dge_data %>%
        filter(umis > umi_cutoff) %>%
        group_by(percentage) %>%
        summarise(median_reads = median(reads),
                  median_umis = median(umis),
                  median_genes = median(genes),
                  median_pcr = median(pcr),
                  num_beads = n()) %>%
    gather('observation', 'value', median_reads:num_beads) %>% 
    as_tibble() %>%
    mutate(percentage = as.integer(percentage),
           observation = factor(observation, levels = c(
                      'median_reads', 
                      'median_umis', 
                      'median_genes',
                      'median_pcr', 'num_beads'))) %>%
    ggplot(aes(percentage, value)) +
        geom_smooth(formula = y~log(x), color = '#56B4E9', size = 0.6) +
        geom_point(size=2, fill = '#56B4E9', color = 'black', pch=21) + 
        scale_x_continuous(breaks=seq(0, 100, 10)) +
        facet_wrap(~observation, scales = 'free_y', ncol=2) +
        labs(y='', x='downsampling percentage')
}
```
\newpage

##### With UMI cutoff 1

```{r plot_umi_1, echo = F, fig.height = 7, fig.width = 7, fig.align='center'}
plot_data(1) %>% print
```

\newpage

##### With UMI cutoff 10

```{r plot_umi_10, echo = F, fig.height = 7, fig.width = 7, fig.align='center'}
plot_data(10) %>% print
```

\newpage

##### With UMI cutoff 50

```{r plot_umi_50, echo = F, fig.height = 7, fig.width = 7, fig.align='center'}
plot_data(50) %>% print
```

\newpage

##### With UMI cutoff 100

```{r plot_umi_100, echo = F, fig.height = 7, fig.width = 7, fig.align='center'}
plot_data(100) %>% print
```
